### PR TITLE
Feature/add audio recording - tested on Windows, Mac, Android

### PR DESCRIPTION
**Pull request by cdavidyoung**
I am at the moment unable to test on iPhone because of some provisioning problem.

I have beefed up the sample app to both record and play and report the duration in each case.

I am sure I could suck the plugin code into my app for more robust testing but it would be nice to import it like a nuget.  Is that easy to do for testing?

Charles

### CODE REVIEW

**Comment by cdavidyoung** [comment]
Actually, I have 5 more commits than is listed.  I guess I have not done the pull request from the right place.

**Comment by cdavidyoung** [comment]
Actually, I see all the commits now.  I'm still getting used to github.

**Comment by bijington** [comment]
@cdavidyoung thanks for this! I will try my best to get the iOS/macOS parts tested

File: samples/AudioPlayerSample/GlobalSuppressions.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+﻿// This file is used by Code Analysis to maintain SuppressMessage
+// attributes that are applied to this project.
+// Project-level suppressions either have no target or are given
+// a specific target and scoped to a namespace, type, member, etc.
+
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("Style", "IDE1006:Naming Styles", Justification = "<Pending>", Scope = "member", Target = "~F:AudioPlayerSample.ViewModels.AudioRecorderPageViewModel._RecordButtonText")]
```
Comment (jfversluis) [inline samples/AudioPlayerSample/GlobalSuppressions.cs:8]: Instead of ignoring these can we fix them? 😊

**Review by jfversluis — CHANGES_REQUESTED** [review]
I didn't go over everything because I think I see a lot of changes that have to do with a different indentation setting in your IDE. This causes complete fils to show up as a change, while only the indentation is changed?

Don't think that is really desirable and makes it a bit hard to review the actual changes here.

**Comment by cdavidyoung** [comment]
Sorry for the inconvenience.  This frequently happens when multiple people
work on files using different editors.  If there had been consistent
indentation to begin with I would not have felt compelled to make the
changes, which consisted of replacing all tabs with spaces, indenting a
consistent 3 spaces, and making the line endings consistent.

I don't like to suppress things either.  I was doing it in this case to
make it easier to substitute for "RecordButtonText" as I was creating new
variables.  I can look at what I can do to fix this.

This suppression comes up a lot and I don't know what to do with it
otherwise:

#pragma warning disable CS1998 // Async method lacks 'await' operators and
will run synchronously

Charles

**Comment by jfversluis** [comment]
> If there had been consistent
indentation to begin with I would not have felt compelled to make the
changes, which consisted of replacing all tabs with spaces, indenting a
consistent 3 spaces, and making the line endings consistent.

If that is the case, that is fine and indeed our bad, but imho should go into a separate PR to mix up all kinds of changes that are now hard to review.

> This suppression comes up a lot and I don't know what to do with it
otherwise:

This is a big topic. Async/await is an important but somewhat complex topic and we definitely shouldn't suppress those. Unfortunately, there isn't an easy answer to what you should do with it. In the easiest form you should turn something like this:

```csharp
private bool DoSomething()
{
   // This triggers CS1998
   DoAnotherSomething();

   return true;
}
```

into an await call:

```csharp
private async bool DoSomething()
{
   await DoAnotherSomething();

   return true;
}
```

But usually it's not this easy unfortunately. There is a lot of info about this here: https://github.com/brminnick/AsyncAwaitBestPractices

**Comment by cdavidyoung** [comment]
Gerald,

I agree with the separate PR going forward.  Before doing any real work I
could do the formatting to get consistent indentation and submit that as a
PR.

Yeah, unfortunately the Async/await issue comes up frequently and can
really clutter the build output window if it is not dealt with.  I
generally try various ways to get rid of it but if all else fails I use the
#pragma.  At least it will be easier to find them all later if a better
solution is found.

#pragma warning disable CS1998 // Async method lacks 'await' operators and
will run synchronously

On a related issue, the error reporting in Maui seems to be more confusing
than it was with Xamarin.  I have gotten used to finding and fixing the
real errors and usually the nonsensical ones go away.  If all else fails
I'll do a rebuild all or just try running and the weird errors magically
disappear.

Charles

**Comment by cdavidyoung** [comment]
I have been able to get rid of all the warnings in my branch without
resorting to suppression.  I have also tested it on the iPhone simulator on
my Mac.

So I'm ready to do a pull request but I can't quite figure it out.  Can one
of you guys help me out here?

Charles

**Comment by bijington** [comment]
@cdavidyoung Good work on the testing of things. Sorry I haven't had a chance to check out the implementation yet but I will hope to do so in the next week or so.

Don't worry this is a pull request so you have done everything that needs to be done there

**Comment by cdavidyoung** [comment]
Shaun, so now that I have done one pull request on my branch I don't need
to do another as I commit changes to my branch?  Can you see the changes
that I made this morning to remove warnings?

Also, I would like to build a library that I can add to my app to do more
thorough testing.  How do I do that before we update the nuget?

Charles

**Comment by cdavidyoung** [comment]
Ok, guys, I would like to move forward with this by doing some more robust
testing in my app.  I figure I could just integrate the code into my
project but I would prefer to do it as a library instead like a nuget.  Is
the latter easy?

Thanks
Charles

**Comment by bijington** [comment]
@cdavidyoung sorry for the delay in getting back to you. In order to include this in your app to test, you can just add the project to your applications solution and then add a reference to the project in your application. That would be the quickest way to achieve it. Basically just replicate what we have done with the sample app but in your application.

**Comment by bijington** [comment]
@cdavidyoung thank you for the effort you have put into this PR. I fear that there are some changes that we won't want to proceed with but I don't want to throw away the work that you have done. I am happy to document the bits we wouldn't want to proceed with here if you feel that would help but I don't want to put any extra pressure on yourself to action it, etc.

I think the best approach might be to merge this PR into the the branch it is targeting and then proceed to tidy some bits up and finish off any testing that we need to do.

**Comment by cdavidyoung** [comment]
Right, well I fully understand that changes are expected as a result of the
review.  As long as it works at least as well as it currently does I will
be happy.

In the meantime I guess I'll just copy the code into my project so that I
can start testing in my app.  I can easily merge changes that come along
and ultimately use the nuget when it is available.

**Comment by cdavidyoung** [comment]
So I notice when I build the solution Plugin.Maui.Audio.sln it gives me a
Plugin.Maui.Audio.dll.  Can I just use that in my app solution instead of
copying over the project folder?

**Comment by cdavidyoung** [comment]
So I tried pulling in the src folder and just got a zillion errors.  So,
once again, is it possible to just link with the Plugin.Maui.Audio.dll?

**Comment by bijington** [comment]
> Right, well I fully understand that changes are expected as a result of the review. As long as it works at least as well as it currently does I will be happy. In the meantime I guess I'll just copy the code into my project so that I can start testing in my app. I can easily merge changes that come along and ultimately use the nuget when it is available.
> […](#)

if you are happy that we work through them here then that is great! I was mostly worried of raising loads of suggestions and leaving you feeling like your work has been wasted. I don't believe it has.

On the topic of it works as it currently does, there will be bits like the extra bits you have added to the interfaces like `Duration()` etc. that I don't think we will keep. Let me add my comments to this PR then and we can see where to go from there

**Comment by bijington** [comment]
> So I tried pulling in the src folder and just got a zillion errors. So, once again, is it possible to just link with the Plugin.Maui.Audio.dll? On Wed, Dec 14, 2022 at 1:56 PM Charles David Young < ***@***.***> wrote:
> […](#)

I am surprised by getting a load of errors and can't really comment without seeing them. You should be able to reference to the `Plugin.Maui.Audio.dll` from your project yes.

**Comment by bijington** [comment]
> So I tried pulling in the src folder and just got a zillion errors. So, once again, is it possible to just link with the Plugin.Maui.Audio.dll? On Wed, Dec 14, 2022 at 1:56 PM Charles David Young < ***@***.***> wrote:
> […](#)

I am surprised by getting a load of errors and can't really comment without seeing them. You should be able to reference to the `Plugin.Maui.Audio.dll` from your project yes.

File: samples/AudioPlayerSample/Platforms/Android/AndroidManifest.xml [hunk 2]
```diff
@@ -3,4 +3,9 @@
 	<application android:allowBackup="true" android:icon="@mipmap/appicon" android:roundIcon="@mipmap/appicon_round" android:supportsRtl="true"></application>
 	<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
 	<uses-permission android:name="android.permission.INTERNET" />
+	<uses-permission android:name="android.permission.RECORD_AUDIO"/>
+	<uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
+	<uses-permission android:name="android.permission.CAMERA" />
```
Comment (bijington) [inline samples/AudioPlayerSample/Platforms/Android/AndroidManifest.xml:8]: Do we need Camera permission here?

File: samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs [hunk 3]
```diff
@@ -4,36 +4,117 @@ namespace AudioPlayerSample.ViewModels;
...
+   Color recordButtonColor = Colors.Blue;
+   public Color RecordButtonColor
+   {
+      get => recordButtonColor;
+      set
+      {
+         recordButtonColor = value;
+         NotifyPropertyChanged();
+      }
+   }
```
Comment (bijington) [inline samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs:34]: I would prefer to stick to firmer MVVM concepts and manage this part in the view/page. We can swap colors/text based on a state or boolean property to know whether the view model is currently recording.

File: samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs [hunk 4]
```diff
@@ -4,36 +4,117 @@ namespace AudioPlayerSample.ViewModels;
...
+   {
+      try
+      {
+         this.page = page;
+      }
+      catch (Exception ex)
+      {
+         page.DisplayAlert("Alert", $"setPage Exception: {ex.Message}", "OK");
+      }
+   }
```
Comment (bijington) [inline samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs:64]: The view model really should never know about the view. If there isn't a fancy interface provided by .NET MAUI in order to call `DisplayAlert` from a view model perhaps we should roll a little `IAlertService` or something

File: samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs [hunk 5]
```diff
@@ -4,36 +4,117 @@ namespace AudioPlayerSample.ViewModels;
...
+   
+   public Command PlayCommand { get; set; }
+   public async void playAudio()
+   {
+      if (audioSource != null)
+      {
+         audioPlayer = this.audioManager.CreatePlayer(((FileAudioSource)audioSource).GetAudioStream());
+         audioPlayer.PlaybackEnded += DonePlaying;
+         await Task.Run(() =>
```
Comment (bijington) [inline samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs:78]: This shouldn't need to be async

File: samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs [hunk 6]
```diff
@@ -4,36 +4,117 @@ namespace AudioPlayerSample.ViewModels;
...
 
-			this.audioManager.CreatePlayer(((FileAudioSource)audioSource).GetAudioStream()).Play();
-		}
-		catch (Exception ex)
-		{
+   public Command StartCommand { get; set; }
+   async void Start()
+   {
+      // This must be done for Android to avoid an exception but I don't think it would hurt in any case
+      if (await HavePermissionMicrophoneAsync())
```
Comment (bijington) [inline samples/AudioPlayerSample/ViewModels/AudioRecorderPageViewModel.cs:89]: Perhaps we should wrap this up into something to demonstrate to consumers on how to check for permission or should we consider adding it into the recording implementation?

File: src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.android.cs [hunk 7]
```diff
@@ -7,198 +7,226 @@ namespace Plugin.Maui.Audio;
...
+         if (duration == 0)
+            duration = player.Duration / 1000.0;
+         return duration;
+      }
+   }
+
+   void t_Tick(object? sender, EventArgs e)
+   {
+      Ts = DateTime.Now - startTime;
+   }
```
Comment (bijington) [inline src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.android.cs:34]: I am not sure I fully follow what this is trying to achieve? Why couldn't we just expose `player.Duration`. If we do go ahead I don't think this is be in the scope of this PR though

File: src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.macios.cs [hunk 8]
```diff
@@ -5,104 +5,134 @@ namespace Plugin.Maui.Audio;
...
+            duration = player.Duration / 1000.0;
+         }
+         return duration;
+      }
+   }
+
+   void t_Tick(object? sender, EventArgs e)
+   {
+      Ts = DateTime.Now - startTime;
+   }
```
Comment (bijington) [inline src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.macios.cs:31]: See my point in the Android code base about what this is trying to achieve.

File: src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.windows.cs [hunk 9]
```diff
@@ -5,139 +5,167 @@ namespace Plugin.Maui.Audio;
...
+         if (duration == 0)
+            duration = player.PlaybackSession.NaturalDuration.TotalSeconds;
+         return duration;
+      }
+   }
+
+   void t_Tick(object? sender, EventArgs e)
+   {
+      ts = DateTime.Now - startTime;
+   }
```
Comment (bijington) [inline src/Plugin.Maui.Audio/AudioPlayer/AudioPlayer.windows.cs:29]: See my point in the Android code base about what this is trying to achieve.

File: src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs [hunk 10]
```diff
@@ -6,23 +6,28 @@ namespace Plugin.Maui.Audio;
...
-    Task<IAudioSource> StopAsync();
+   ///<Summary>
+   /// Stop recording and return the AudioRecording instance with the recording data
+   ///</Summary>
+   Task<IAudioSource> StopAsync();
+
+   ///<Summary>
+   /// Returns the duration of the last audio recording or playback.
+   ///</Summary>
+   double Duration();
```
Comment (bijington) [inline src/Plugin.Maui.Audio/AudioRecorder/IAudioRecorder.shared.cs:32]: If anything this should live on the `IAudioSource` interface

**Comment by cdavidyoung** [comment]
Shaun, thanks for moving this forward!  I will restrict my responses to
your comments about the plugin.

So the Duration is something that I need in my app and I figure it might be
useful for others as well.  However, I could live without it and just
implement the timer code in my app.

Meanwhile, I am looking into how to use the dll directly.

Thanks
Charles

**Comment by cdavidyoung** [comment]
I am initially testing in Windows so I tried replacing
Plugin.Maui.Audio.dll in the
C:\Users\charl\source\repos\GitHub\jiffyLog\bin\Debug\net7.0-windows10.0.19041.0\win10-x64
and Apx folders.  However, when I run the app in the debugger the original
dll appears again.  I'll continue to look for ways to replace the dll that
stick.

**Comment by cdavidyoung** [comment]
I have been able to test the Plugin.Maui.Audio code on Windows and Android
by replacing

C:\Users\charl\.nuget\packages\plugin.maui.audio\0.1.0-preview1\lib\net6.0-windows10.0.19041\
Plugin.Maui.Audio.dll
C:\Users\charl\.nuget\packages\plugin.maui.audio\0.1.0-preview1\lib\net6.0-android31.0\Plugin.Maui.Audio.dll

with their corresponding dll files from the plugin build.  The only hiccup
is that VS 2022 Preview reports errors like

Error CS0246 The type or namespace name 'IAudioRecorder' could not be found
(are you missing a using directive or an assembly reference?) jiffyLog
(net7.0-ios), jiffyLog (net7.0-maccatalyst)
C:\Users\charl\source\repos\GitHub\jiffyLog\ViewModels\AudioService.cs 10
Active

However, it builds and runs fine even though it is reporting a number of
errors.

When I tried to do the same thing on the Mac side for MacCatalyst and iOS
the errors are fatal and it won't let me run the app.

The bottom line is that I have been able to now test the dlls in two
platforms, which gives me a good idea they will run fine on the other
platforms, especially because they have been tested with the sample app.  I
am not using the GetDuration() call so feel free to remove it or if you
would like I can do it (along with the Timer code) and push the changes.

Let me know if there is anything else I can do and I hope we can get this
nuget published soon!
Charles

**Comment by cdavidyoung** [comment]
I finally found the easy way to test this package locally:

***@***.***/creating-and-using-a-local-nuget-package-repository-9f19475d6af8

so now I have tested on all 4 platforms.  The only change I had to make to
the project file was to enable the Generate Nuget package on build option.

Charles

**Comment by cdavidyoung** [comment]
I am not sure why the nuget has not been released.  I ended up building the nuget locally from my branch.  If you look up a couple of messages you will see (what was) a link to a site that explains how to create and load a local nuget.

https://github.com/cdavidyoung/Plugin.Maui.Audio/tree/feature/add-audio-recording

**Comment by cdavidyoung** [comment]
https://github.com/justinsmits, you will want to try the sample app in my branch as well to make sure it works for you. It will record and playback snippets on all 4 platforms.

**Comment by strnbrgd** [comment]
When you use ISO AVAudioRecorder and not specify file extionsion, it will use apple's caff format to wrap your audio, making it very difticult to play on other platforms.

Need to put an extension according to your preferred AudioFormatType so it will not use it (In my case - aac was not playable on Windows Media Player or Google Nest until this change).

**Comment by cdavidyoung** [comment]
I can't really address the specifics of your comment, @strnbrgd.  I can only say that I have tested this plugin on all 4 platforms and have not noticed any problem playing audio recorded on one platform on any other platform.  The audio comes in as .wav and I convert it to .mp3 with ffmpeg.

**Comment by strnbrgd** [comment]
When recording on IOS, the file comes as .tmp file.
When changed to aac format it saved as .tmp file which could only be opened in windows via VLC and no other player.
When opening it you can see it starts with caff (Apple's Core Audio Format) and inner format aac.
After changing the extension in the url passed to the AVAudioRecorder, the file was created without caff container, so it could be player on any toher platform **without** any conversion.
My goal is to play the recorded file in google nest so I couldn't afford any conversion

**Comment by o0JoeCool0o** [comment]
Need support for recording audio cant believe this isnt in yet!! How can I implement this PR in my project? it sounds like its working? can I create a custom version of the Maui framework that uses this somehow?

**Comment by jfversluis** [comment]
@o0JoeCool0o please dial it down a little. This is a community effort which we put time in outside of our regular work. 

Features will come and be merged as we have time. If you need this sooner, you’re free to roll your own implementation.

**Comment by o0JoeCool0o** [comment]
Dialed down... Appreciate the community effort! I was surprised this wasn't included by Microsoft as native functionality for their MAUI.NET Platform. Apologies for the outburst, Ive been working long hours to get a feature implemented only to get slapped by having audio playable but not recordable. I am trying this solution now to roll my own. https://www.syncfusion.com/blogs/post/building-an-audio-recorder-and-player-app-in-net-maui.aspx

 But was hoping there was an easy way to bring these changes into my project to reduce these long hours :)

**Comment by jfversluis** [comment]
Totally understandable and that is why we are working hard to make your life easier! 

.NET MAUI is primarily a UI framework, so it isn’t that strange that something like this isn’t part of the package imo. But still great to be able to have it and use it easily. If you have anything that you found in your long hours, please feel free to add something to this PR, thanks!

**Comment by cdavidyoung** [comment]
@o0JoeCool0o, if you don't mind building the nuget and installing it in your project this has been working for me on all 4 platforms:
https://github.com/cdavidyoung/Plugin.Maui.Audio/tree/feature/add-audio-recording

**Comment by bijington** [comment]
This will get merged at some point in the not so distant future. I have taken some of @cdavidyoung work that he kindly provided and am trying to mould it into our desired API surface. Sadly that has proven challenging plus a lot of personal stuff has got in the way

File: src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.windows.cs [hunk 11]
```diff
@@ -1,25 +1,30 @@
...
 using Windows.Media.MediaProperties;
 using Windows.Storage;
 
 namespace Plugin.Maui.Audio;
-
+
 partial class AudioRecorder : IAudioRecorder
 {
-	MediaCapture mediaCapture;
+    MediaCapture? mediaCapture;
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.windows.cs:9]:
```suggestion
	MediaCapture? mediaCapture;
```

File: src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.windows.cs [hunk 12]
```diff
@@ -1,25 +1,30 @@
...
+	string audioFilePath = string.Empty;
 
 	public bool CanRecordAudio { get; private set; } = true;
-
-	public bool IsRecording => mediaCapture != null;
-
-	string audioFilePath;
+    public bool IsRecording => mediaCapture != null;
```
Comment (jfversluis) [inline src/Plugin.Maui.Audio/AudioRecorder/AudioRecorder.windows.cs:13]:
```suggestion
	public bool IsRecording => mediaCapture != null;
```

**Review by jfversluis — COMMENTED** [review]
Couple of very small things. I think overall we're pretty good here!

File: samples/AudioPlayerSample/AudioPlayerSample.csproj [hunk 13]
```diff
@@ -28,8 +28,14 @@
...
 		<TargetPlatformMinVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">10.0.17763.0</TargetPlatformMinVersion>
 		<SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'tizen'">6.5</SupportedOSPlatformVersion>
+		<ProvisioningType>automatic</ProvisioningType>
 	</PropertyGroup>
 
+	<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net6.0-ios|AnyCPU'">
+	  <CreatePackage>false</CreatePackage>
+	  <CodesignProvision>Automatic</CodesignProvision>
+	  <CodesignKey>iPhone Developer</CodesignKey>
+	</PropertyGroup>
```
Comment (jfversluis) [inline samples/AudioPlayerSample/AudioPlayerSample.csproj:38]: Probably also remove this, although same story as the other: will be added back by VS probably...

File: samples/AudioPlayerSample/GlobalSuppressions.cs (hunk 1 above)
Comment (bijington) [reply samples/AudioPlayerSample/GlobalSuppressions.cs]: Oh yeah how did that creep in?
